File: assignment2/python/task4.py
# coding=UTF-8
import numpy as np
import color
import utils
import noise
import time
import stats


def main():
	image = utils.load_image('lenna.png')
	grayscale_image = color.rbg2gray(image)
	noise_images = []

	for i in range(0, 15):
		noise_images.append(noise.salt_pepper(grayscale_image, 0.1))

	noise_images = np.array(noise_images)
	
	t = time.clock()
	
	mean_image = _multiple_images_processing(noise_images, mean)
	
	tt = time.clock()
	print "Mean time %s." % (tt-t)

	median_image = _multiple_images_processing(noise_images, stats.median)
	
	tt = time.clock()
	print "Median time %s." % (tt-t)
	
	utils.display_single_image('Original', grayscale_image)
	utils.display_single_image('Noise sample',  noise_images[0])
	utils.display_single_image('Mean result', mean_image)
	utils.display_single_image('Median result', median_image)

    #utils.save_image('yiq.png', yiq)

	utils.wait_key_and_destroy_windows()

def mean(ls):
	sum = 0
	for i in ls:
		sum=sum+i
	return sum/(len(ls))

def _multiple_images_processing(images, f):	
	result = np.empty_like(images[0,:,:])

	for i in range(0, images.shape[1]):
		for j in range(0, images.shape[2]):
			pixels = images[:,i,j]
			result[i,j] = f(pixels)	

	return result


if __name__ == "__main__":
	main()

File: assignment2/matlab/task4.m
% Produza uma pilha com 15 imagens da mesma paisagem com ruídos diferentes. Pode-se
% gerar um ruído aleatório para uma única imagem, utilizar uma câmera que tire várias
% fotos seguidas ou converter os frames de um vídeo de câmera estática em imagens.
% Aplique:
%   a. Filtragem por Média de Imagens;
%   b. Filtragem por Mediana de Imagens.
function task4() 
    assetsDir = ['..' filesep 'assets' filesep];
    filePath = strcat(assetsDir, 'lenna.png');
    originalImage = imread(filePath);
        
    yiqImage = rgb2ntsc(originalImage);
    % Get grayscale image
    grayIm = yiqImage(:,:,1);
    
    numIm = 15;
    noiseImages = createNoiseImages(grayIm, numIm, 0.2);        
            
    meanImage = mean(noiseImages,3);
    meanDiff = imDiff(meanImage, grayIm);
    
    medianImage = median(noiseImages,3);
    medianDiff = imDiff(medianImage, grayIm);
        
    % Ploting
    fig = figure(1);    
    set (fig, 'Units', 'normalized', 'Position', [0,0,1,1]);
    
    subplot(2,2,1), imshow(grayIm), title('Original image')
    subplot(2,2,2), imshow(noiseImages(:,:,1)), title('Image Noise Sample')  
    subplot(2,2,3), imshow(meanImage), title(strcat('Mean Filter. Err: ',num2str(meanDiff*100,'%5.1f%%')))
    subplot(2,2,4), imshow(medianImage), title(strcat('Median Filter. Err: ',num2str(medianDiff*100,'%5.1f%%')))
          
end

function output = createNoiseImages(img, n, p)
    output = [];    
    for i = 1:n
        im = imnoise(img,'salt & pepper',p);        
        output = cat(3,output, im);        
    end
    
end

% Calculates difference between two images in percentage
function output = imDiff(I, J)    
    imDiff = I-J; 
    imSum = I+J;  
    output = 200 * mean(imDiff(:)) / mean(imSum(:));
end